### PR TITLE
[FLINK-6346] Migrate from Java serialization for GenericWriteAheadSink's state

### PR DESCRIPTION
**Pull request by fanyon**
fix Migrate from Java serialization for GenericWriteAheadSink's state

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java [hunk 1]
```diff
@@ -88,15 +93,23 @@ public void initializeState(StateInitializationContext context) throws Exception
 		Preconditions.checkState(this.checkpointedState == null,
 			"The reader state has already been initialized.");
 
-		checkpointedState = context.getOperatorStateStore()
-			.getSerializableListState("pending-checkpoints");
+		javaCheckpointedState = context.getOperatorStateStore().getSerializableListState(checkpointedStateDescriptor.getName());
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java]: It's not necessary to store this in a field; it's only used within this method anyway.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java [hunk 2]
```diff
@@ -88,15 +93,23 @@ public void initializeState(StateInitializationContext context) throws Exception
...
 		if (context.isRestored()) {
 			LOG.info("Restoring state for the GenericWriteAheadSink (taskIdx={}).", subtaskIdx);
 
-			for (PendingCheckpoint pendingCheckpoint : checkpointedState.get()) {
-				this.pendingCheckpoints.add(pendingCheckpoint);
+			try {
+				for (PendingCheckpoint pendingCheckpoint : javaCheckpointedState.get()) {
+					this.pendingCheckpoints.add(pendingCheckpoint);
+				}
+			} catch (Exception e) {
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java]: What exceptions are we expecting here?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java (hunk 1 above)
Comment (zhangminglei) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java]: @zentol I think he stores it in this field for later to write in ```snapshotState ```.

**Comment by zhangminglei** [comment]
@fanyon Another suggestion here. Make the lots of commit log into one log is nice.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java (hunk 1 above)
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java]: It's never references anywhere outside this method though. `checkpointState` is, but not `javaCheckpointState`.

Comment (zhangminglei) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java]: Just remove ```javaCheckpointedState``` filed is ok for this purpose.

Comment (zhangminglei) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java]: @zentol I just think he forgot it. So, like that.

**Comment by zentol** [comment]
For clarification: Commits should **not** be squashed once the PR has been opened. However, *before* opening a PR it would be good to squash these minor commits.

**Comment by zhangminglei** [comment]
```git push upstream -f xxx ``` DOES NOT WORK for squashed it ?

**Comment by zentol** [comment]
@zhangminglei I don't understand what you're trying to say.

**Comment by zhangminglei** [comment]
@zentol Sorry. I just drunk. My mean is we can use ```git rebase -i origin/master``` to combine lots of commit logs into one log and then use ```git push origin -f branch_issueId``` push to orgin server.

**Comment by zentol** [comment]
There is no doubt that you **can** modify the commits of an open PR. My point is that you **shouldn't** because it messes with the review.

This applies to both squashing commits and force pushing in general. The only exception is when you rebase the PR; but that should only be done if a rebase actually causes conflicts.

**Comment by fanyon** [comment]
@zhangminglei @zentol thanks for your suggestions, and I have updated the code. 

As discussed in [https://github.com/apache/flink/pull/3750#issuecomment-298869900](url), it may not need to register two state like the current way, and older version of the data could be converted in some way, such as upgrade tools. I look forward to hearing your thoughts after you think about it later. @tzulitai

**Comment by tzulitai** [comment]
Hi, sorry for the late response here @fanyon.
I've left a comment on the umbrella JIRA on how we proceed with this. Could you take a look? Thanks :-D

**Pull request closed without merging**